### PR TITLE
Apply workspace.exclude to workspace.default-members.

### LINKED ISSUE #8460: `workplace.default-members` and `workspace.exclude` don't always work well together.
One of the listed use cases for `workspace.exclude` in the [documentation](https://doc.rust-lang.org/cargo/reference/workspaces.html) is for when you are "using a glob pattern and you want to remove a directory".

This is very handy, and lets `workspace.members` stay pretty clean if your project structure is *mostly* consistent. Unfortunately it isn't similarly supported by `workspace.default-members`, which *must* be a strict subset of `workspace.members`. 

To see what I mean, here's an example I hit recently (this is based on the project layout of https://github.com/mozilla/application-services -- but simplified).

Essentially, I think something like the following code should work:

```toml
[workspace]
members = [
    "components/*",
    "components/support/*",
    "tools/i-compile-very-slowly-and-am-rarely-needed",
]

# Exclude a couple things that the globs in `workspace.members` catch
# overeagerly
exclude = [
    # This is not a crate, and just contains support and utility code.
    "components/support",
    # Android support code (this is in Kotlin, not Rust)
    "components/support/android",
]

# Everything except `tools/i-compile-very-slowly-and-am-rarely-needed`
default-members = [
    "components/*",
    "components/support/*",
]
```

However, if you try this, you get errors like "`<project>/components/support/android` is listed in workspace’s default-members but is not a member", because "When specified, default-members must expand to a subset of members."

### Possible solution

I think that rule makes sense, and can help catch bugs, so I'm not saying it should be removed or weakened. I think my issue can be fixed by just tweaking the way that rule is implemented rather than removing it:

Instead of checking that `workspace.default-members` is a subset of `(workspace.members - workspace.exclude)` (which is what cargo does now), cargo should check that `default-members` is a subset of `members`.

Conceptually, I think this is really just a change of *when* the check happens -- e.g. IMO it should happen prior to applying exclude to members, as opposed to afterwards.

(Of course, I think you will also need to apply `workspace.exclude` to `workspace.default-members` to produce the computed set of default members, but that's easy and if the above holds, it shouldn't remove anything from it except what was removed from `workspace.members`)

Anyway, thanks for your time.

---

P.S. Sorry that this doesn't fit an issue template. I wrote it out in a text editor first to make sure it made sense, but now it's hard to adapt and I don't really have time to rewrite it. Let me know if you need any clarification.

Also, since it's kinda halfway between bug and feature request, the template didn't really made tons of sense, I think.

I'm also sorry that the issue title is not very descriptive.

### PR DESCRIPTION
**Pull request by thomcc**
Not sure how controversial the feature request was, it seemed easy to do, so I did it (I'm aware it's possible this won't be accepted).

Fixes #8460

### CODE REVIEW

**Comment by thomcc** [comment]
This actually doesn't go all the way to fixing my issue, which I believe would also require fixing https://github.com/rust-lang/cargo/issues/6745. It's not clear to me which part of this code is causing that, but if someone has a concrete idea I could take a stab at that too.

File: src/cargo/core/workspace.rs [hunk 1]
```diff
@@ -538,14 +538,26 @@ impl<'cfg> Workspace<'cfg> {
...
+                let normalized_path = paths::normalize_path(&path);
+                let manifest_path = normalized_path.join("Cargo.toml");
                 if !self.members.contains(&manifest_path) {
+                    // default-members are allowed to be excluded, but they
+                    // still must be referred to by the original (unfiltered)
+                    // members list. Note that we aren't testing against the
+                    // manifest path, both because `members_paths` doesn't
+                    // include `/Cargo.toml`, and because excluded paths may not
+                    // be crates.
+                    let exclude = members_paths.contains(&normalized_path)
```
Comment (alexcrichton) [inline src/cargo/core/workspace.rs:556]: How come this condition is here as well? I may not be fully remembering what the `members_paths` set is either...

File: tests/testsuite/workspaces.rs [hunk 2]
```diff
@@ -1628,6 +1628,94 @@ fn exclude_but_also_depend() {
...
+        .file("bar/quux/Cargo.toml", &basic_manifest("quux", "0.1.0"))
+        .file("bar/quux/src/lib.rs", "");
+
+    let p = p.build();
+    p.cargo("build").run();
+
+    assert!(p.root().join("target").is_dir());
+    assert!(!p.bin("foo").is_file());
+    assert!(p.bin("baz").is_file());
+
```
Comment (ehuss) [inline tests/testsuite/workspaces.rs:1683]: Maybe also add a check `!p.bin("quux").exists()` here and the `--workspace` case?

**Comment by ehuss** [comment]
If you want to work on #6745, I think the issue is the use of `starts_with` on [this line](https://github.com/rust-lang/cargo/blob/2a64ae86b9ad003beb81decadf5b556a0ca63059/src/cargo/core/workspace.rs#L1165-L1176). I'm uncertain why it is using starts_with, but I think it should be something more like `manifest_path.parent().unwrap() == self.root_dir.join(mem)`.  

I personally have a hard time following the workspace member discovery code, so it's a bit tricky to know for sure what to change, but that part definitely doesn't look right to me.

File: src/cargo/core/workspace.rs (hunk 1 above)
Comment (thomcc) [reply src/cargo/core/workspace.rs:556]: `members_paths` is the glob-expanded paths of `members` (without `Cargo.toml`s). Without this check I believe the test (Which this patch adds) `excluded_default_members_still_must_be_members` failed -- e.g. you could have `members = ["foo"]`, `default-members = ["foo", "bar"]` and `exclude = ["bar"]`.

File: tests/testsuite/workspaces.rs (hunk 2 above)
Comment (thomcc) [reply tests/testsuite/workspaces.rs:1684]: Done.

**Comment by thomcc** [comment]
> If this doesn't end up actually solving your issue though that's sort of unfortunate :(. I think fixing that though sounds reasonable to add here too!

If possible, I think I'd rather try and fix #6745 as a separate issue -- This doesn't entirely fix my issue but it does make things slightly better, and it might be a week or two before I have a chance to dig into the other issue (it's also not as immediately obvious to me as this one was -- as mentioned, this code is a little confusing).

**Pull request merged**